[H1410101] iP

H1410101 opened the pull request:
_For class reasons; ignore._

# CatBot

You might be wondering, why would you use CatBot?

- it has cats

1. There is no better reason, and there needs no other reason.

It features peak programmer syntax, like [this amazing, lovable ~~monstrosity~~ little guy](https://github.com/H1410101/ip/blob/master/src/main/java/catbot/CatBot.java). 

Here's a snippet for your convenience:
```java
BiConsumer<String, BiFunction<
                NamedParameterMap, BiConsumer<ErrorIndicatorIo.InvalidArgumentState, NamedParameterMap>,
                Optional<Task>>>
                createTaskIfValidElseWarn = (args, bifunction) -> slashPattern.ifParsableElseDefault(args,
                        namedParameterMap -> bifunction.apply(
                                namedParameterMap,
                                io::indicateArgumentInvalid
                        ).ifPresent(task -> {
                            taskList.addTask(task);
                            io.displayTaskAdded(taskList);
                        })
                );

        commands.addCommand("todo",
                        args -> createTaskIfValidElseWarn.accept(args, Task.Todo::createIfValidElse)
                )
                .addCommand("event",
                        args -> createTaskIfValidElseWarn.accept(args, Task.Event::createIfValidElse)
                )
                .addCommand("deadline",
                        args -> createTaskIfValidElseWarn.accept(args, Task.Deadline::createIfValidElse)
            );
```

**That's right!** Your :eyes: do not deceive you. That is a 
- [x] `BiConsumer` that takes in a 
- [x] `BiFunction`, which takes in a 
- [x] `BiConsumer` as one of its arguments
- [x] AND returns an `Optional`! 

I bet you've never seen that many angle brackets in a row since CS2030S.

> Programs must be written for machines to execute, and only incidentally for people to partially comprehend in flashes of brilliance after large swathes of agony. 
~ Me, 2023, while descending into the infinite abyss, 2 weeks behind schedule

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

H1410101 committed:
Feature: greet and farewell

H1410101 committed:
Feature: echo until bye

H1410101 committed:
Feature: task types, task doneness, task list, and parser for slash arguments

H1410101 committed:
Feature: textuitesting

H1410101 committed:
Feature: handled various errors, implemented exceptions for missing argument errors

H1410101 committed:
Feature: delete

H1410101 committed:
Restructure project (transient)

H1410101 committed:
Completely restructures

H1410101 committed:
Allow saving of todo list information through serialization

H1410101 committed:
Merge branch 'branch-Level-7'

H1410101 committed:
Add support for dates

H1410101 committed:
Merge branch 'branch-Level-8'

rayyan35p commented on src/main/java/catbot/CatBot.java:
```diff
@@ -0,0 +1,102 @@
+                )
+                .addCommand("unmark",
+                        string -> runIfValidIndexElseIndicateError.accept(string,
+                                validIndex -> {
+                                    taskList.unmarkTask(validIndex-1);
+                                    io.printTaskModified(taskList, validIndex);
+                                }
+                        )
+                )
+                .addCommand("delete",
+                        string -> runIfValidIndexElseIndicateError.accept(string,
+                                validIndex -> io.printTaskDeleted(taskList.removeTask(validIndex-1))
+                        )
+                );
+
+        // User creating new tasks (with SlashPattern)
+
+        BiConsumer<String, BiFunction<NamedParameterMap, BiConsumer<ErrorIndicatorIo.InvalidState, NamedParameterMap>, Optional<Task>>>
+                createTaskIfValidElseWarn = (args, bifunction) -> slashPattern.ifParsableElseDefault(args,
+                        namedParameterMap -> bifunction.apply(
+                                namedParameterMap,
+                                io::indicateArgumentInvalid
+                        ).ifPresent(task -> {
+                            taskList.addTask(task);
+                            io.printTaskAdded(taskList);
+                        })
+                );
+
+        commands.addCommand("todo",
+                        args -> createTaskIfValidElseWarn.accept(args, Task.Todo::createIfValidElse)
+                )
+                .addCommand("event",
+                        args -> createTaskIfValidElseWarn.accept(args, Task.Event::createIfValidElse)
+                )
+                .addCommand("deadline",
+                        args -> createTaskIfValidElseWarn.accept(args, Task.Deadline::createIfValidElse)
+                )
+        ;
+
+    }
```
Perhaps you could have added this to another class like CommandMap?

H1410101 replied:
I actually spent quite a bit deliberating this, but I ended up not doing it (at least for now).

## The part where I defend myself
My thought process was that I wanted to add bot-specific things in bot-specific classes; to change what commands are supported and what behaviours result from commands, my current iteration only requires that I change or even replace `CatBot` (and possibly `CatBotConsoleIO`) with minimal rewrite.
While there might be some merit in a `CatBotCommands` class that contains this method, that would leave CatBot as a small, anaemic class (I am leaning towards this being a bad thing, but I'm not too sure of myself here).
Ultimately, I had decided that it made more sense to me to have the commands be clear from the bot description, instead of its own separate class, especially since the class would otherwise be so small as to be trivial.

## The part where I stop defending myself
I definitely see an alternative that takes your considerations into account. I could rename the current file to `CatBotMain` or `CatBotEntrypoint` to reflect its explicit purpose as a small and trivial class, put this initialization as a protected static method of `CatBotCommandLoader`, or have it be the default initialization of `CatBotCommandMap`. 
I'd love to hear your thoughts on why this design might be preferable, though; or if you had a different design in mind, I'd like to chew on that as well.

Regardless, thanks for the feedback!

rayyan35p commented on src/main/java/catbot/CatBotCommandPatterns.java:
```diff
@@ -0,0 +1,63 @@
+
+import java.util.function.Consumer;
+
+public class CatBotCommandPatterns {
```
Maybe you can consider adding header comments/javadocs? For all your other classes as well

H1410101 replied:
Yup! I'll get around to it. _Eventually_. :')

rayyan35p commented on src/main/java/catbot/Parser.java:
```diff
@@ -0,0 +1,80 @@
+     * @param map to store the mapping between command and argument
+     */
+    private static void parseCmdArg(String s, NamedParameterMap map) {
+        String[] cmdArg = s.split("\\s", 2);
```
Perhaps you can write out the variable name of `cmdArg` to be in full so that it is clearer?

H1410101 replied:
Good catch, will do. Thanks!

rayyan35p commented on src/main/java/catbot/Parser.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    private static void parseCmdArg(String s, NamedParameterMap map, boolean emptyArgument) {
+        if (emptyArgument) {
+            map.addNamedParameter("", s.trim());
+        } else {
+            parseCmdArg(s, map);
+        }
+    }
```
I like that you still wrapped in curly brackets even though they are single statement!

rayyan35p reviewed (commented):
Good code quality mostly, just need to add header comments and rename variables to make it clear. 💯

remuslum commented on src/main/java/catbot/internal/NamedParameterMap.java:
```diff
@@ -1,11 +1,37 @@
+        parameters = new HashMap<>();
+    }
+
+    public boolean containsKey(String key) {
```
can consider renaming method to avoid confusion

remuslum reviewed (commented):
Good job 💯

nubnubyas commented on src/main/java/catbot/CatBot.java:
```diff
@@ -0,0 +1,102 @@
+
+        // User creating new tasks (with SlashPattern)
+
+        BiConsumer<String, BiFunction<NamedParameterMap, BiConsumer<ErrorIndicatorIo.InvalidState, NamedParameterMap>, Optional<Task>>>
```
Perhaps u can split this line into 2? I think this line has 135 characters which exceeds the 120 characters limit stated in the coding standard.

H1410101 replied:
Got it, and thanks for all the nudges for coding style. 
I ended up running it through a stylechecker (finally), but I appreciate the pointers!

nubnubyas commented on src/main/java/catbot/io/CatbotConsoleIO.java:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    public String getCommand() {
+        if (this.endOfCommand == null) findCommand();
+        if (this.endOfCommand == -1) return retrieve();
+        else return retrieve().substring(0, this.endOfCommand);
+    }
+
+    public String getArgs() {
+        if (this.endOfCommand == null) findCommand();
+        if (this.endOfCommand == -1) return "";
+        else return retrieve().substring(this.endOfCommand+1);
```
Perhaps it would be better to have the conditionals on a separate line. E.g
if (this.endOfCommand == null) {
    findCommand();
}

nubnubyas commented on src/main/java/catbot/io/CatbotConsoleIO.java:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    private void findCommand() {
+        if (retrieve() != null)
+            this.endOfCommand = retrieve().indexOf(" ");
```
It would be better to have curly brackets here based on the java coding standard.

nubnubyas reviewed (commented):
Good job for the chatbot! I really like your usage of java.util.function in your code. Just some small coding standard problems here and there. Overall, I really like your IP.

btaanish commented on src/main/java/catbot/CatBot.java:
```diff
@@ -0,0 +1,102 @@
+                .generate(io::indicateInvalidInteger);
+        CommandPattern<NamedParameterMap> slashPattern = CatBotCommandPatterns.getSlashPatternGenerator()
+                .generate(CatBotCommandPatterns.NO_DEFAULT);
+
```
I like how you name the variables in a greater explanatory sense

btaanish commented on src/main/java/catbot/Parser.java:
```diff
@@ -0,0 +1,80 @@
+        return by(delimiter, false);
+    }
+
+    public static Parser by(String delimiter, boolean emptyArgument) {
```
Perhaps adding javadocs for public functions might help explain their functioning in greater detail?

H1410101 replied:
Noted! I'll probably get back to it at some point.

btaanish commented on src/main/java/catbot/CatBotCommandPatterns.java:
```diff
@@ -0,0 +1,63 @@
+    //endregion
+
+    //region Slash Arguments Pattern
+    private static final SlashArgumentPatternGenerator slashPatternGenerator = new SlashArgumentPatternGenerator();
```
Perhaps having shorter variable names may increase the readability of the code?

H1410101 replied:
Perhaps, but these are very infrequently used, so I thought it might have been better to make it very descriptive instead.
I'll think about it, but as of right now I'm still leaning towards my current convention, even if it results in excessively long variable names.

btaanish commented on src/main/java/catbot/io/CatbotConsoleIO.java:
```diff
@@ -0,0 +1,229 @@
+        return s.replaceAll("(^|\n)", "$1"+prefix);
+    }
+
+    private void line() {
```
Perhaps adding a verb in the function name might be more intuitive?

H1410101 replied:
Good catch; I'll do that!

btaanish commented on src/main/java/catbot/task/TaskList.java:
```diff
@@ -0,0 +1,108 @@
+        return removed;
+    }
+
+    public void ifValidIndexElse(int index, Consumer<Integer> ifValid, Consumer<Integer> otherwise) {
```
Perhaps adding context to the variable names might communicate their purpose better?

H1410101 replied:
This is heavily inspired by `Optional`'s [`ifPresentOrElse`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Optional.html#ifPresentOrElse(java.util.function.Consumer,java.lang.Runnable)) (and other similar methods), and I'm trying to take it to its logical extensions; if, even with context, this does not seem intuitive enough, I'm happy to discuss about how to make it better!

btaanish reviewed (commented):
Overall a nice attempt with a close adherence to the coding standars. Just a few nits here and there but generally well designed.

H1410101 committed:
Rename for clarity and convention

H1410101 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

H1410101 committed:
Verify Gradle and JAR

H1410101 committed:
Add JUnit tests and fix minor bugs

H1410101 committed:
Add JavaDoc to half of the classes, especially Interfaces

H1410101 committed:
Fix all coding standard except JavaDoc

H1410101 committed:
Add find functionality

H1410101 committed:
Merge branch 'branch-Level-9'

H1410101 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/catbot/CatBotCommandPatterns.java

H1410101 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	Tasks.txt
#	src/main/java/catbot/CatBot.java
#	src/main/java/catbot/CatBotCommandPatterns.java
#	src/main/java/catbot/io/CatbotConsoleIO.java
#	src/main/java/catbot/task/Task.java
#	src/main/java/catbot/task/TaskList.java

H1410101 committed:
Refactor for better OOP and style

H1410101 committed:
Refactored for OOP, fixed indexing bug

H1410101 committed:
Add working JavaFX prototype, expecting refactoring and reorganisation

H1410101 committed:
Add working JavaFX UI for chatbot

H1410101 committed:
Merge branch 'branch-A-Varargs'

# Conflicts:
#	Tasks.txt
#	src/main/java/catbot/bot/CatBot.java

H1410101 committed:
Merge branch 'branch-A-CheckStyle'

# Conflicts:
#	src/main/java/catbot/io/CatBotConsoleIo.java
#	src/main/java/catbot/io/UserIo.java
#	src/main/java/catbot/task/TaskList.java

H1410101 committed:
Add javafxplugin to build.gradle

H1410101 committed:
Remove irrelevant todos and clean up warnings

H1410101 committed:
Add update/edit functionality

H1410101 committed:
Fix style for checkstyle

Mainly added javadocs for classes, and their constructors. Also included
javadocs for all public functions.
Private or protected functions were not thoroughly documented.

Test classes were not fixed, and not updated for the new functionality.

H1410101 committed:
Fix non-checkstyle code quality

Went in to be more thorough with code quality.
I already wrote most of the things with code quality in mind, so this is
not a very huge change.
Added comments where I deliberately diverged from general code quality
advice.

H1410101 committed:
Merge pull request #2 from H1410101/branch-A-Assertions

Assertions, but it's actually an extension.

H1410101 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/catbot/io/CatBotConsoleIo.java

H1410101 committed:
Merge pull request #3

Code Quality, but I actually didn't do much

H1410101 committed:
Make header comments adhere to conventions

Especially the linguistic form of the starting word, which I did not pay
attention to while coding.

H1410101 committed:
Update GUI, implemented delayed application exit

H1410101 committed:
Add UI screenshot, add acknowledgement when task list is empty

H1410101 committed:
Update README.md

H1410101 committed:
Added acknowledgement when unnumbered lists are empty

H1410101 committed:
Merge remote-tracking branch 'origin/master'

H1410101 committed:
Polish UI, including rescaling images, changing padding, etc

H1410101 committed:
Format README.md
